File: src/main/java/rip/hippo/mosey/util/IOUtil.java
/*
 * Mosey is a free and open source java bytecode obfuscator.
 *     Copyright (C) 2020  Hippo
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rip.hippo.mosey.util;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

/**
 * @author Hippo
 * @version 1.0.0, 6/23/20
 * @since 1.0.0
 */
public enum IOUtil {
    ;

    public static byte[] toByteArray(JarFile jarFile, JarEntry jarEntry) throws IOException {
        InputStream inputStream = jarFile.getInputStream(jarEntry);
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        int length = (int)(jarEntry.getSize() / 8);
        byte[] buffer = new byte[length];
        for(int len = inputStream.read(buffer); len > 0; len = inputStream.read(buffer)) {
            byteArrayOutputStream.write(buffer, 0, len);
        }
        return byteArrayOutputStream.toByteArray();
    }
}

File: src/main/scala/rip/hippo/mosey/util/IOUtil.scala
package rip.hippo.mosey.util

import java.io.ByteArrayOutputStream
import scala.annotation.tailrec
import java.util.jar.{JarEntry, JarFile}

/**
 * @author Hippo
 * @version 1.0.0, 7/10/20
 * @since 1.0.0
 */
object IOUtil {

  def toByteArray(jarFile: JarFile, jarEntry: JarEntry): Array[Byte] = {
    val inputStream = jarFile.getInputStream(jarEntry)
    val byteArrayOutputStream = new ByteArrayOutputStream()
    val length = (jarEntry.getSize / 8).asInstanceOf[Int]
    val buffer = new Array[Byte](length)
    @tailrec def read(): Unit = {
      val len = inputStream.read(buffer)
      if (len > 0) {
        byteArrayOutputStream.write(buffer, 0, len)
        read()
      }
    }
    read()
    byteArrayOutputStream.toByteArray
  }
}
